src/sign/interactive_sign/participant.rs

```diff
@@ -0,0 +1,164 @@
+///     one for each participant (including this participant), and
+///   - A single set of [private auxiliary
+///     information](`crate::auxinfo::AuxInfoPrivate`) for this participant.
+///
+///
+/// # Protocol output
+/// Upon successful completion, the participant outputs a [`Signature`].
+/// The signature is on the message which was used to produce the provided
+/// input message digest. It verifies under the public verification key
+/// defined by the [keygen output](crate::keygen::Output).
+///
+///
+/// [^cite]: Ran Canetti, Rosario Gennaro, Steven Goldfeder, Nikolaos
+/// Makriyannis, and Udi Peled. UC Non-Interactive, Proactive, Threshold ECDSA
+/// with Identifiable Aborts. [EPrint archive,
+/// 2021](https://eprint.iacr.org/2021/060.pdf).
+#[derive(Debug)]
+pub struct InteractiveSignParticipant {
+    input: Input,
+    presigner: PresignParticipant,
+    signer: SignParticipant,
+}
+
+/// Input for the interactive signing protocol.
+#[derive(Debug)]
+#[allow(unused)]
+pub struct Input {
+    message_digest: Sha256,
+    presign_input: presign::Input,
+}
+
+#[derive(Debug, PartialEq, Eq)]
+pub enum Status {
+    /// Participant is created but has not received a ready message from self.
+    NotReady,
+    /// Participant received a ready message and is running presign.
+    RunningPresign,
+    /// Participant completed presign and is running sign.
+    RunningSign,
+    /// Participant completed signing and output a signature.
+    TerminatedSuccessfully,
+}
+
+impl ProtocolParticipant for InteractiveSignParticipant {
+    type Input = Input;
+    type Output = Signature;
+
+    type Status = Status;
+
+    fn ready_type() -> MessageType {
+        PresignParticipant::ready_type()
+    }
+
+    fn protocol_type() -> ProtocolType {
+        todo!()
+    }
+
+    #[allow(unused)]
+    fn new(
+        sid: Identifier,
+        id: ParticipantIdentifier,
+        other_participant_ids: Vec<ParticipantIdentifier>,
+        input: Self::Input,
+    ) -> Result<Self> {
+        todo!()
+    }
+
+    fn id(&self) -> ParticipantIdentifier {
+        // Note: signer should have the same participant ID
+        self.presigner.id()
+    }
+
+    fn other_ids(&self) -> &[ParticipantIdentifier] {
+        // Note: signer should have the same set of other IDs
+        self.presigner.other_ids()
+    }
+
+    #[allow(unused)]
+    fn process_message<R: RngCore + CryptoRng>(
+        &mut self,
+        rng: &mut R,
+        message: &Message,
+    ) -> Result<ProcessOutcome<Self::Output>> {
+        todo!()
+    }
+
+    fn status(&self) -> &Self::Status {
+        // This method makes some assumptions about ordering for calling presign
+        // and sign -- e.g. we will not pass a ready message to the `signer` until
+        // the `presigner` is sucessfully completed.
+        // Another option would be to maintain a status field and update it at
+        // the appropriate poitns.
```

Comment:
@gatoWololo I'd be interested to hear your thoughts on this. I'm slightly starting to move this implementation to having a bit more state machine mechanics explicitly, rather than the current "hope the right things are in storage" approach. But I'm not doing it in a very principled way.

Reply:
Right, that makes sense. Do you mean my thoughts on adding state machine states to the different implementors of `ProtocolParticipant`s or the ordering assumptions for the `SignParticipant` and `PresignParticipant` here?

Either way, I agree keeping track of the state is better than relying on storage. Random thought:
- I don't know whether you can use the same `Status` enum across different `*Participant`s (I guess not since other participants probably want to keep track of the rounds). That's fine though, the enums are lightweight and understandable. So its fine if we have a few floating around.
- The way you have currently implemented `status`, even with the assumptions, seems good to me as well.

Reply:
Cool, thanks. I think we can discuss best practices for this more if we try to use statuses more consistently, but glad you don't see any red flags with this.